### PR TITLE
New docker flow including pulling images

### LINKED ISSUE #611: New docker flow
The goal is to publish FPC images to accelerate development. To achieve this I suggest the following change of our docker flow.

# FPC docker 

## Use cases

The docker images provided in this repository target different use cases:

1) *Used as FPC dev environment*
   
   Provides a complete development environment including Intel SGX compiler, etc... 
   Useful to develop and test FPC locally without installing FPC dependencies on local dev machine. 
   This refers to preferred option to getting started with FPC.

1) *Used as CI environment*
   
   Based on FPC dev environment and used during Github CI. 

1) *Used as runtime environment for FPC components (FPC chaincode and ERCC)*
   
   Required to package, deploy, and run FPC chaincode. 

## Docker Images

FPC comes with the following docker images to build FPC components and the development environment.
All images start with `hyperledger/fabric-private-chaincode-` prefix.

* `base-rt`: Base image for FPC. Includes all runtime dependencies including SGX runtime services.
* `ccenv`: Chaincode environment image for FPC chaincodes based on `base-rt`.
* `base-dev`: Base development image. Includes all build tools for FPC including protobuf, SGX SSL, SGX compiler, ...
* `dev`: Development image based on `base-dev`. Add additional user-defined tools and development dependencies.

These images can be build manually (see [Building images](#building-images) section) or pulled from `ghcr.io/hyperledger/fabric-private-chaincode` (see [Pulling images](#pulling-images) section).

## Building images

* `make build`: creates base-rt, ccenv
  
* `make build-dev`: creates base-rt, base-dev
  

## Pulling images

* `make pull`: pulls ccenv

* `make pull-dev`: pulls base-dev 

Note that base-rt not pulled as it is an intermediate image included already as layer in `ccenv` and `base-dev`.

## Running FPC dev environment

* `make run-dev`: creates dev (if not exist) and runs it. Does not create base-dev and base-rt, returns an error if not exists.

### Start docker-based FPC dev environment

Option 1) Pull images and start dev container

```bash
cd utils/docker
make pull-dev
make run-dev
## continue inside docker
```

Option 2) Build images from scratch and start dev container
```bash
cd utils/docker
make build-dev
make run-dev
## continue inside docker
```

### CI

If no changes in `utils/docker`
```bash
cd utils/docker
make pull
make run-dev
## continue inside docker
```

otherwise
```bash
cd utils/docker
make build-dev
make run-dev
## continue inside docker
```

### Build docker images for FPC runtime environment

Pull images
```bash
cd utils/docker
make pull
```

or build them
```bash
cd utils/docker
make build
```

## Publishing

We publish images for every new release (docker tag = `${FPC_VERSION}`) and on PR merged (docker tag = main) through CI.
TBD: Instead of pushing to main, we could also introduce a nightly build?

* `make publish`:
  pushing base-rt, ccenv, base-dev to github docker registry 

When running the `publish` target manually (not through CI), docker login on `ghcr.io` is needed.
See [documentation](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry).

Use your github personal access token (PAT) to login.
```bash
export CR_PAT=YOUR_TOKEN
echo $CR_PAT | docker login ghcr.io -u USERNAME --password-stdin
```

TBD: Let's only publish ubuntu 20.04 LTS based images.

### PR DESCRIPTION
**Pull request by mbrandenburger**
This PR introduces changes the docker flow as discussed in #611.

FPC docker images can now be pulled from github registry or built manually.

Closes #611

### CODE REVIEW

**Comment by ikegawa-koshi** [comment]
@mbrandenburger 
I just tried to run FPC following the steps in this PR.
I was able to pull that docker image without any problem.
It works correctly in my environment (Azure SGX HW).
Very nice work!

**Comment by mbrandenburger** [comment]
removed automated publishing from this PR. This will come with a followup PR.

TODO: `DOCKER_REGISTRY ?= ghcr.io/mbrandenburger` needs to be changed to `DOCKER_REGISTRY ?= ghcr.io` before merging

**Comment by munapower** [comment]
I was able to create the development environment with both methods provided. When I tried to follow the test-network sample I ran into this error:
```make -C $FPC_PATH/utils/docker pull
make: Entering directory '/project/src/github.com/hyperledger/fabric-private-chaincode/utils/docker'
# ccenv
DOCKER_BUILDKIT=1 docker  pull ghcr.io/mbrandenburger/hyperledger/fabric-private-chaincode-ccenv:main
Error response from daemon: Head https://ghcr.io/v2/mbrandenburger/hyperledger/fabric-private-chaincode-ccenv/manifests/main: unauthorized
make: *** [Makefile:242: pull] Error 1
make: Leaving directory '/project/src/github.com/hyperledger/fabric-private-chaincode/utils/docker'```

**Comment by mbrandenburger** [comment]
```
Error response from daemon: Head https://ghcr.io/v2/mbrandenburger/hyperledger/fabric-private-chaincode-ccenv/manifests/main: unauthorized
```

Ouch that was my fault. I did not hit the public button on github for this image. It should work now.

**Comment by munapower** [comment]
> ```
> Error response from daemon: Head https://ghcr.io/v2/mbrandenburger/hyperledger/fabric-private-chaincode-ccenv/manifests/main: unauthorized
> ```
> 
> Ouch that was my fault. I did not hit the public button on github for this image. It should work now.

now I got stopped in the make build in test-network

``` make build
make -C /Users/mmunaro/go/src/github.com/hyperledger-labs/fabric-private-chaincode/ercc all docker
# hard to list explicitly, so just leave empty target,
# which forces ecc to always be built
# ERCC's binary is created here.
# The binary is then referenced by the fpc-peer cli, inserted in the ERCC package,
# and eventually run by the external launcher.
go  build -tags WITH_PDO_CRYPTO -o ercc main.go
# github.com/hyperledger/fabric-private-chaincode/ercc/attestation
ld: unknown option: --start-group
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
# github.com/hyperledger/fabric-private-chaincode/internal/utils
ld: unknown option: --start-group
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [ercc] Error 2
make: *** [ercc-container] Error 2
```

**Comment by munapower** [comment]
@mbrandenburger I was able to follow de readmes for FPC dev environment and the sample test-network both pulling as well as building the images.

File: utils/docker/dev/Dockerfile
Comment (bvavala) [inline utils/docker/dev/Dockerfile:3]: I want to make the point that we should remove this `dev` image, and leave the `base-dev` one. This requires minor modifications, because both images can `build` FPC, though only `dev` can run all the FPC tests.

The main difference between `base-dev` and `dev`  is the presence of Fabric's source code and the `SGX_MODE` variable in `dev`.
The former, in particular, does not change, since we check out a predefined version.
So we are leaving to the user to do the job (while building the `dev` image), but we can do this before. Additionally, Fabric's code is necessary only because of some tests under `internal`. Otherwise, even `base-dev` can run the FPC tests.
Anyway, without opening a discussion on the tests, I would leave the source code in the image.

The `SGX_MODE` variable instead is set/hardcoded in the image. However, the `dev` image is meant to be used in SIM/HW mode. So, this should not be hardcoded -- besides, we already have `SIM` as default value.
Additionally, any such variable can be set at docker run-time (with the `-e` option). Hence, there is no need to create an image just for setting it. Actually, it would be even desirable to set it at run-time, equal to the value of the host's variable. 

Concluding, `dev` could be removed, and this would not impact the build process. However, for our tests (including local tests in HW mode) to work, `base-dev` must check out Fabric's code, and `make run-dev` (or equivalent target for running `base-dev`) must set the `SGX_MODE` variable appropriately.

**Review by bvavala — COMMENTED** [review]
Minor comments and one suggestion to remove `dev`.

File: utils/docker/dev/Dockerfile
Comment (mbrandenburger) [reply utils/docker/dev/Dockerfile:3]: I think you made a good point. In general I vote for removing the dev image but I still see a good use for it.

Regarding `SGX_MODE` I agree with you and we should set it at docker run-time.

Regarding Fabric, this can also go away once tests are "refactored".

Let me add another difference between `base-dev` and `dev`. The `dev` image may contain custom applications installed via `apt`. The user may set `DOCKER_DEV_IMAGE_APT_ADD_PKGS` to install whatever they need inside the dev container. For instance, I use this feature to install a few things, like `tmux` and `vim` in my dev container, which I will not miss!! :P

Moving `DOCKER_DEV_IMAGE_APT_ADD_PKGS` frpm `dev` to `base-dev` would require to rebuild it from scratch and not leveraging the new docker pull mode. Works but not nice when you just want to have `tmux` installed.
Installing these packages on every startup also not ideal as the container is not persistent.
Let the user create a "custom" image based on `base-dev` that contains additional tools they want, would lead to the solution we already have with dev.

File: utils/docker/Makefile
Comment (mbrandenburger) [reply utils/docker/Makefile:132]: not exactly, `$(FPC_DOCKER_NAMESPACE)-dev` is the image name, whereas `$(DOCKER_DEV_CONTAINER_NAME)` is the container name. For whatever reasons they are different.  

But you made a point that `$(FPC_DOCKER_NAMESPACE)-dev` can be replaced with `$(FPC_DOCKER_DEV_NAME)` in the `docker run` command.

File: utils/docker/dev/Dockerfile
Comment (mbrandenburger) [reply utils/docker/dev/Dockerfile:24]: a left over :)

File: utils/docker/Makefile
Comment (bvavala) [reply utils/docker/Makefile:134]: that's right -- I have confused the variables.

File: utils/docker/dev/Dockerfile
Comment (bvavala) [reply utils/docker/dev/Dockerfile:3]: > Regarding Fabric, this can also go away once tests are "refactored".

"Away", when tests are refactored -- I guess you know better -- or "in base-dev" at any time to consolidate the two images.

> The user may set DOCKER_DEV_IMAGE_APT_ADD_PKGS ... 

True, but that's not something that impacts FPC: neither the build process, nor the testing (assuming Fabric's code is present, of course).
For this reason, the question is whether FPC should provide this feature, or leave this to the end users/developers. Expectedly, if they need anything other than an apt package, they _can_ and will build their own images, but FPC will not provide any such feature/enablement in dev.
Anyway, fine to keep this for now to avoid any disruption. I have captured this in #616 .

**Review by bvavala — APPROVED** [review]
lg, tested in hw mode

final step before merge: switch the ghcr link to the hyperledger namespace

**Comment by mbrandenburger** [comment]
@bvavala I changed the namespace and pushed the images to 
- https://github.com/orgs/hyperledger/packages/container/package/fabric-private-chaincode-ccenv
- https://github.com/orgs/hyperledger/packages/container/package/fabric-private-chaincode-base-dev

Please check again that you can pull the images before (squashing) merging. Thank you!

**Pull request merged**
